### ISSUE TITLE
Record landing page: record statistics frontend implementation

### ISSUE THREAD

**Issue by kpsherva** [inveniosoftware/invenio-stats#120]
1. The design should take into account currently provided stats:

![](https://user-images.githubusercontent.com/1814661/175041762-a0901d49-d472-4e24-b70a-b203a87b28f4.png)

2. The design should take into account the confusion of the users on how statistcs are computed (once per day). There should be clearly visible information
3. The design should also apply to record search view

Comments:
  - Seems more reasonable to display the Unique Views first (check the latest spec of COUNTER and see what is the recommendation there)
  - Explain in the UI that admins/curators/owners are excluded (if implemented)

Nice to have:
Graph visualisations with time evolution

Prerequisites:
- finalised backend payload or finalised backend

# Decisions

## Mockups

![216320834-2d6acb55-cf83-4256-b0b4-61dc8be45f8e](https://user-images.githubusercontent.com/22594783/217199903-f940ac92-3fd1-4314-a1ac-6d315db5c122.png)

## [DISCUSSION NEEDED] Event payload for aggregated views (mock response)

- [ ] We need to decide how the backend response will be to incorporate all the statistics needed to be shown. Below is a sample response but fields can vary.

### OpenSearch: Events

Record view:
```json
    "hits": [
      {
        "_index": "my-site-events-stats-record-view-2023-02-14",
        "_id": "2023-02-14T10:44:20-9650b423ae31a0fadd60c51d58c38b55f3956eb1",
        "_score": 1,
        "_source": {
          "timestamp": "2023-02-14T10:44:26",
          "referrer": "https://127.0.0.1:5000/",
          "recid": "g0k00-37e25",
          "parent_recid": "hbf9a-8sr53",
          "via_api": false,
          "is_robot": false,
          "country": null,
          "visitor_id": "a0465c28b0b8144c3da4673cd9734668fd860a26bf91221125cb171e",
          "unique_session_id": "d3342f948c9774c5fad255c9d2afd01c01753b292b898493bfea2fa6",
          "unique_id": "ui_g0k00-37e25"
        }
      },
```

File download:
```json
      {
        "_index": "my-site-events-stats-file-download-2023-02-14",
        "_id": "2023-02-14T10:44:30-50be4e8552b18f262724d7ce401eb25991189f02",
        "_score": 1,
        "_source": {
          "timestamp": "2023-02-14T10:44:30",
          "bucket_id": "222b7b2d-9f21-4b2b-9199-bb630cea6ee2",
          "file_id": "0e359078-1379-4b61-85a5-f6af0a988c75",
          "file_key": "62kyt6(2).jpg",
          "size": 49457,
          "referrer": "https://127.0.0.1:5000/records/g0k00-37e25",
          "recid": "g0k00-37e25",
          "parent_recid": "hbf9a-8sr53",
          "via_api": true,
          "is_robot": false,
          "country": null,
          "visitor_id": "a0465c28b0b8144c3da4673cd9734668fd860a26bf91221125cb171e",
          "unique_session_id": "d3342f948c9774c5fad255c9d2afd01c01753b292b898493bfea2fa6",
          "unique_id": "222b7b2d-9f21-4b2b-9199-bb630cea6ee2_0e359078-1379-4b61-85a5-f6af0a988c75"
        }
      },
```

### OpenSearch: Aggregations

Record views:
```json
    "hits": [
      {
        "_index": "my-site-stats-record-view-2023-02",
        "_id": "ui_g0k00-37e25-2023-02-14",
        "_score": 1,
        "_source": {
          "timestamp": "2023-02-14T00:00:00",
          "unique_id": "ui_g0k00-37e25",
          "count": 2,
          "unique_count": 1,
          "recid": "g0k00-37e25",
          "parent_recid": "hbf9a-8sr53",
          "via_api": false
        }
      }
    ]

 ```

Record downloads (all file downloads for a record):
```json
    "hits": [
      {
        "_index": "my-site-stats-file-download-2023-02",
        "_id": "222b7b2d-9f21-4b2b-9199-bb630cea6ee2_0e359078-1379-4b61-85a5-f6af0a988c75-2023-02-14",
        "_score": 1,
        "_source": {
          "timestamp": "2023-02-14T00:00:00",
          "unique_id": "222b7b2d-9f21-4b2b-9199-bb630cea6ee2_0e359078-1379-4b61-85a5-f6af0a988c75",
          "count": 1,
          "unique_count": 1,
          "volume": 49457,
          "file_id": "0e359078-1379-4b61-85a5-f6af0a988c75",
          "file_key": "62kyt6(2).jpg",
          "bucket_id": "222b7b2d-9f21-4b2b-9199-bb630cea6ee2",
          "recid": "g0k00-37e25",
          "parent_recid": "hbf9a-8sr53"
        }
      }
    ]
```

### Python Invenio-Stats query results

Record views:
```python
{'start_date': None,
 'end_date': None,
 'recid': 'g0k00-37e25',
 'parent_recid': 'hbf9a-8sr53',
 'count': 2.0,
 'unique_count': 1.0}
```

Record downloads:
```python
{'start_date': None,
 'end_date': None,
 'recid': 'g0k00-37e25',
 'parent_recid': 'hbf9a-8sr53',
 'count': 2.0,
 'unique_count': 2.0,
 'volume': 51339.0}
```

**Comment by jennur** [comment]
## Statistics segment

- [SUI styled accordion](https://react.semantic-ui.com/modules/accordion/#types-styled)
- [SUI definition table](https://react.semantic-ui.com/collections/table/#types-definition)
- [SUI info icon message](https://react.semantic-ui.com/collections/message)

<img width="421" alt="Screenshot 2023-01-25 at 16 44 28" src="https://user-images.githubusercontent.com/21052053/214609679-6173e256-36e3-4753-8462-6843b9b9b7f7.png">
<img width="421" alt="Screenshot 2023-01-25 at 16 49 13" src="https://user-images.githubusercontent.com/21052053/214609672-2064e11e-bda3-4387-8672-cde059d3c853.png">


## On landing page
![statistics_collapsed](https://user-images.githubusercontent.com/21052053/214608293-72e38d4d-2aa2-46d2-bc0c-aa015505af1a.png)
![statistics_open](https://user-images.githubusercontent.com/21052053/214608310-99b0e1d4-cd77-4fec-9ec0-157ef1778756.png)

**Comment by kpsherva** [comment]
for the search view, I have found some old mockups, how about placing the stats as it suggests? see:

![Image](https://user-images.githubusercontent.com/38131488/215430849-8f6e6494-1396-4476-b17a-e90413bc40eb.png)

**Comment by jennur** [comment]
> for the search view, I have found some old mockups, how about placing the stats as it suggests? see:
> 
> ![Image](https://user-images.githubusercontent.com/38131488/215430849-8f6e6494-1396-4476-b17a-e90413bc40eb.png)

## Search list item v2
- [SUI tiny label](https://react.semantic-ui.com/elements/label/)
![Screenshot 2023-01-30 at 10 45 10](https://user-images.githubusercontent.com/21052053/215442796-8dd8f1ff-d1eb-401e-815e-dd5d0c501122.png)

**Comment by jennur** [comment]
## Conclusions (discussed irl)

- Keep the unstyled accordion
- "Read more" on new line
- Recharts is preferred (fits better with SUI)
- Test search list item for request list as well (where we also have buttons on top)
- Split task in to (statistics component, graph section[nice to have])

**Comment by jennur** [comment]
## Updated layout

### Final statistics layout after discussion
- [SUI statistic](https://react.semantic-ui.com/views/statistic/#content-value)
    - Label icons: view, download
- [SUI accordion](https://react.semantic-ui.com/modules/accordion/#types-standard)
- [SUI definition table](https://react.semantic-ui.com/collections/table/#types-definition)
- [SUI info icon message](https://react.semantic-ui.com/collections/message)

<img width="400" src="https://user-images.githubusercontent.com/21052053/215714128-1a3de38f-16ee-40cc-bae1-7ae7b7559e85.png"/>
<img width="400" src="https://user-images.githubusercontent.com/21052053/215714131-dee9cf38-b51e-435c-a5de-a3ccfe3c2c04.png"/>

### Nice to have (separate task)

- Graph by [Recharts](https://recharts.org/en-US)
<img width="400" src="https://user-images.githubusercontent.com/21052053/215714125-3da54865-8185-4ee7-a073-022069e1e3c8.png"/>

<img width="400" src="https://user-images.githubusercontent.com/21052053/215715421-a1de46be-ed63-4c69-8d91-bacfe2d4277c.png"/>

**Comment by jennur** [comment]
## Search list item v3

### Front page
![Screenshot 2023-01-31 at 10 08 37](https://user-images.githubusercontent.com/21052053/215716971-da10ffb1-b762-4eec-b403-6dca8603cdb0.png)

### My dashboard
![Screenshot 2023-01-31 at 10 04 47](https://user-images.githubusercontent.com/21052053/215716972-539a8115-79e9-42aa-8994-eb74d7f8a3a7.png)

**Comment by lnielsen** [comment]
The big blue box "All statistics updated daily" seems to big and scary to me...

**Comment by kpsherva** [comment]
to test: 6,7 figures statistics

**Comment by ntarocco** [comment]
Nice work! A few comments on my side:
- The table's titles/label "Unique views", "All versions" maybe look too bold? The important data is the values.
- Warning about using React for graphs (if we will do it at some point): prefer pure JS if possible. To be discussed if this will be implemented.

**Comment by jennur** [comment]
More options for the info-box:

- No icon:
![Screenshot 2023-02-02 at 10 16 37](https://user-images.githubusercontent.com/21052053/216282329-d270c3d0-6257-454a-9612-afb668411c05.png)

- No icon, tiny text
![Screenshot 2023-02-02 at 10 18 52](https://user-images.githubusercontent.com/21052053/216282875-c6a1c46e-5ec5-4665-9176-b06951be2006.png)


- Gray (but this is a bit against the consistency we have with blue info boxes)
![Screenshot 2023-02-02 at 10 12 32](https://user-images.githubusercontent.com/21052053/216282576-591031f3-dc0c-4900-9b98-090b8cb53ba1.png)


For the table titles, it's the default SUI styling. (We also have bold titles on e.g. the file table in the deposit form.) If we unbold it, I feel like it gets slightly harder to relate the cell to each heading, as the headings don't stand out? WDYT?
![Screenshot 2023-02-02 at 10 26 45](https://user-images.githubusercontent.com/21052053/216284918-736b782c-344d-411d-be51-46bc597a6570.png)
vs
![Screenshot 2023-02-02 at 10 22 28](https://user-images.githubusercontent.com/21052053/216283753-909511d7-48c5-4b47-92a1-55b1109f6ab8.png)

**Comment by jennur** [comment]
For the graphs, [Chart.js](https://www.chartjs.org/) is not React, so we could use that one instead of Recharts, or I maybe [CancasJS](https://canvasjs.com/docs/charts/chart-types/html5-line-chart/) could work instead?

**Comment by jennur** [comment]
**Bigger numbers**

![Screenshot 2023-02-02 at 10 55 25](https://user-images.githubusercontent.com/21052053/216292547-32e12f52-580c-45b4-a4c9-96e45473248a.png)

**Comment by ntarocco** [comment]
> For the table titles, it's the default SUI styling. (We also have bold titles on e.g. the file table in the deposit form.) If we unbold it, I feel like it gets slightly harder to relate the cell to each heading, as the headings don't stand out? WDYT?

I prefer it "unbold", the grey background does the job, but I think it is a personal choice, so it would be nice to have other opinions.

> Bigger numbers
I am again the annoying guy, sorry! Often, when the numbers are very high, others display a more readable version and then, on hover or on click with an icon or something, display the real number, e.g.:
`1.4 M` on hover/click, `1,490,811`. Like the Tb for the file size.
But again it is just IMO, a preference.

When it will implemented, do not forget to take into account the [internationalization](https://www.w3.org/International/wiki/Locale-based_forms) ;)

**Comment by jennur** [comment]
![Screenshot 2023-02-02 at 13 09 10](https://user-images.githubusercontent.com/21052053/216320834-2d6acb55-cf83-4256-b0b4-61dc8be45f8e.png)
![Screenshot 2023-02-02 at 13 08 44](https://user-images.githubusercontent.com/21052053/216320837-57699e73-97ed-4826-99a2-363048d8a93d.png)
I like the idea with the "1.4M"! The text also works good like this I think! The unbold headings I'm still not sure, I agree it maybe looks cleaner, but I feel like it's getting harder to quickly connect the cells when reading it... More opinions would be nice indeed! 😄

**Comment by kpsherva** [comment]
I like the 1.4M, looks very good and it scales up nicely. 
I would remove the bold formatting from the headers, numbers are more important than headings, but by highlighting them we actually give them more importance.

**Comment by jennur** [comment]
To elaborate on my argument... I actually disagree that the number is the most important, as the number doesn't make sense without the headings, which makes it important to being able to connect the cells easily. As there are two headings they are even more difficult to connect. It's not a huge difference, but a slight one.

**Comment by max-moser** [comment]
# Info: using the stats REST API endpoint

The Invenio-Stats module provides a REST API endpoint where the configured queries can be called with a custom set of parameters. Example:

```bash
mmoser@Alpha ~/rdm/my-site $ curl -ks -X POST -H "authorization: bearer $API_TOKEN" \
-H "content-type: application/json" https://localhost:5000/api/stats -d @query.json
{
  "downloads": {
    "count": 5.0,
    "end_date": null,
    "parent_recid": "s24n2-pkw51",
    "recid": "5j71c-mr205",
    "start_date": null,
    "unique_count": 3.0,
    "volume": 3836161.0
  },
  "downloads-all": {
    "count": 6.0,
    "end_date": null,
    "parent_recid": "s24n2-pkw51",
    "start_date": null,
    "unique_count": 4.0,
    "volume": 4705401.0
  },
  "views": {
    "count": 3.0,
    "end_date": null,
    "parent_recid": "s24n2-pkw51",
    "recid": "5j71c-mr205",
    "start_date": null,
    "unique_count": 1.0
  },
  "views-all": {
    "count": 4.0,
    "end_date": null,
    "parent_recid": "s24n2-pkw51",
    "start_date": null,
    "unique_count": 2.0
  }
}
```

With the following `query.json`:
```json
{
    "views": {
        "stat": "record-view",
        "params": {
            "recid": "5j71c-mr205"
        }
    },
    "views-all": {
        "stat": "record-view-all-versions",
        "params": {
            "parent_recid": "s24n2-pkw51"
        }
    },
    "downloads": {
        "stat": "record-download",
        "params": {
            "recid": "5j71c-mr205"
        }
    },
    "downloads-all": {
        "stat": "record-download-all-versions",
        "params": {
            "parent_recid": "s24n2-pkw51"
        }
    }
}
```

As can be seen above, multiple queries can be performed with a single API call.
The top-level keys (`views`, `views-all`, `downloads`, and `downloads-all`) are arbitrary names for the queries and can be chosen freely.
For each query, the `stat` value defines which configured query to call and must refer to one of the names in `current_app.config["STATS_QUERIES"]`.
The `params` will be supplied as keyword arguments to the query's `q.run(...)` method, which will obviously influence the results.
In our example, reasonable `params` would be: `start_date`, `end_date`, and `recid` (or `parent_recid` for the `*-all-versions` variants).

**Comment by lnielsen** [comment]
@max-moser Does the stats endpoint take access control into account? I.e. can I get the number of downloads for a restricted record that I don't have access to?

**Comment by lnielsen** [comment]
It looks great but I have questions and comments obviously :-)

![216320837-57699e73-97ed-4826-99a2-363048d8a93d](https://user-images.githubusercontent.com/1698163/222205817-c7289f87-c581-4a50-aeb7-0e836942e32b.png)

**Comment by ntarocco** [comment]
> It looks great but I have questions and comments obviously :-)
> 
> ![216320837-57699e73-97ed-4826-99a2-363048d8a93d](https://user-images.githubusercontent.com/1698163/222205817-c7289f87-c581-4a50-aeb7-0e836942e32b.png)

@kpsherva Lars comments to be discussed all together :)

**Comment by max-moser** [comment]
@lnielsen I've added a very simple "deny all" policy on the statistics API endpoint (as opposed to the default "allow all" from `invenio-stats`), so there's no leaking of anything.
This means that the statistics are currently only available in the UI, but it shouldn't be too hard to fix up that policy s.t. it resolves the records and checks if the _current identity_ has access to the (parent) record IDs specified for the queries.

In the current form, the statistics are fetched directly in the backend (via `query` objects from `invenio-stats` in the view function) and passed to the jinja templates.
If you don't have read access to the record, the view function won't be executed fully and thus you're not gonna get the statistics for records that you can't access.

The exception here is if there's a record with at least one restricted and one public **version** – the queries (and their results) don't take visibility of the "sibling" records into account and just give you the views for _this version_, as well as _all versions_.
In case there's exactly one restricted and one public version of the same record, figuring out how many views/downloads "the other version" has is a matter of simple subtraction.

**TL;DR:** Currently we only give out statistics on the UI landing pages, and there we do leak a bit of information about sibling versions.
